From 533df4960e0937a4ceceefd0d6a3282838a760ac Mon Sep 17 00:00:00 2001
From: Stephan Gerhold <stephan.gerhold@linaro.org>
Date: Tue, 5 Aug 2025 17:44:32 +0200
Subject: [PATCH 09/22] remoteproc: qcom_q6v5: Send SMP2P stop signal in
 attached/detached state

If we stop the q6v5 remoteproc while it is in RPROC_ATTACHED or
RPROC_DETACHED state, we still want to send the stop signal to shut it down
cleanly.

The main goal of the check in qcom_q6v5_request_stop() is to avoid sending
duplicate shutdown/stop signals during crash or shutdown via sysmon, so
check for RPROC_CRASHED to handle all of RPROC_RUNNING, RPROC_ATTACHED and
RPROC_DETACHED. It does not make sense to call the function while in
RPROC_OFFLINE state.

Signed-off-by: Stephan Gerhold <stephan.gerhold@linaro.org>
---
 drivers/remoteproc/qcom_q6v5.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/drivers/remoteproc/qcom_q6v5.c b/drivers/remoteproc/qcom_q6v5.c
index 3c173a4fc..ed1b578a8 100644
--- a/drivers/remoteproc/qcom_q6v5.c
+++ b/drivers/remoteproc/qcom_q6v5.c
@@ -227,13 +227,12 @@ static irqreturn_t q6v5_stop_interrupt(int irq, void *data)
  */
 int qcom_q6v5_request_stop(struct qcom_q6v5 *q6v5, struct qcom_sysmon *sysmon)
 {
-	bool was_running = q6v5->running;
 	int ret;
 
 	q6v5->running = false;
 
-	/* A watchdog/fatal IRQ clears running; logical crashes still need a stop. */
-	if (!was_running || qcom_sysmon_shutdown_acked(sysmon))
+	/* Don't perform SMP2P dance if remote crashed or already stopped */
+	if (q6v5->rproc->state == RPROC_CRASHED || qcom_sysmon_shutdown_acked(sysmon))
 		return 0;
 
 	qcom_smem_state_update_bits(q6v5->state,
-- 
2.55.0

